common/flatpak-dir.c

```diff
@@ -4018,19 +4022,231 @@ repo_get_remote_collection_id (OstreeRepo *repo,
+  char *token_buffer = NULL;
+  gsize token_size;
+  g_autoptr(GFile) cache_dir = NULL;
+  g_autoptr(GFile) token_cache_file = NULL;
+  g_autoptr(GFileInfo) token_cache_file_info = NULL;
+  GTimeVal token_cache_mtime;
+  GTimeVal current_time;
+  g_autoptr(GMainContext) context = NULL;
+  g_autoptr(SoupRequestHTTP) request = NULL;
+  g_autoptr(SoupMessage) http_message = NULL;
+  g_autoptr(GAsyncResult) request_send_result = NULL;
+  g_autoptr(GInputStream) request_stream = NULL;
+  g_autoptr(GAsyncResult) request_read_all_result = NULL;
+
+  if (api_server_url != NULL)
+    {
+      access_token = get_api_server_access_token (cancellable, error);
+      if (!access_token)
+        {
+          g_prefix_error (error, "Unable to get API server access token: ");
+          return NULL;
+        }
+
+      token_url = g_strconcat (api_server_url,
+                               "/token?ref=",
+                               ref_to_fetch,
+                               "&repo_url=",
+                               remote_url,
+                               NULL);
+
+      token_url_checksum = g_compute_checksum_for_string (G_CHECKSUM_SHA1,
+                                                          token_url,
+                                                          strlen (token_url));
+
+      cache_dir = flatpak_ensure_system_user_cache_dir_location (error);
+      if (!cache_dir)
+        {
+          g_prefix_error (error, "Unable to get system cache dir: ");
+          return NULL;
+        }
+
+      token_cache_filename = g_strconcat (g_file_get_path (cache_dir),
+                                          "/api-server-token-",
+                                          token_url_checksum,
+                                          NULL);
+      token_cache_file = g_file_new_for_path (token_cache_filename);
+      token_cache_file_info = g_file_query_info (token_cache_file,
+                                                 G_FILE_ATTRIBUTE_TIME_MODIFIED,
+                                                 G_FILE_QUERY_INFO_NONE,
+                                                 cancellable, NULL);
+      g_file_info_get_modification_time (token_cache_file_info, &token_cache_mtime);
+      g_get_current_time (&current_time);
+
+      /* Tokens expires in one hour, we won't reuse tokens that are close
+       * to expire, that's why we check for at most 3400 seconds old tokens */
+      if (g_file_query_exists (token_cache_file, cancellable) &&
+          current_time.tv_sec - token_cache_mtime.tv_sec < 3400)
+        {
+          if (!g_file_load_contents (token_cache_file,
+                                     cancellable,
+                                     &token_buffer,
+                                     NULL, NULL,
+                                     error))
+            {
+              g_warning ("Unable to read token cache file %s: %s",
+                         token_cache_filename,
+                         (*error)->message);
```

Comment:
This will crash if the caller passes `NULL` to the `GError **error` argument of `get_repo_bearer_token()`. Use a `g_autoptr(GError) local_error` instead, and use `g_propagate_error (error, g_steal_pointer (&local_error))` to propagate that error to the caller when necessary.